barretenberg/cpp/src/barretenberg/api/api_client_ivc.hpp

```diff
@@ -0,0 +1,432 @@
+#pragma once
+
+#include "barretenberg/api/acir_format_getters.hpp"
+#include "barretenberg/api/api.hpp"
+#include "barretenberg/api/file_io.hpp"
+#include "barretenberg/api/init_srs.hpp"
+#include "barretenberg/api/log.hpp"
+#include "barretenberg/client_ivc/mock_circuit_producer.hpp"
+#include "barretenberg/common/map.hpp"
+#include "barretenberg/common/throw_or_abort.hpp"
+#include "barretenberg/dsl/acir_format/ivc_recursion_constraint.hpp"
+#include "libdeflate.h"
+
+namespace bb {
+
+template <typename T> std::shared_ptr<T> read_to_shared_ptr(const std::filesystem::path& path)
+{
+    return std::make_shared<T>(from_buffer<T>(read_file(path)));
+};
+
+// TODO(#7371): this could probably be more idiomatic
+template <typename T> T unpack_from_file(const std::filesystem::path& filename)
+{
+    std::ifstream fin;
+    fin.open(filename, std::ios::ate | std::ios::binary);
+    if (!fin.is_open()) {
+        throw std::invalid_argument("file not found");
+    }
+    if (fin.tellg() == -1) {
+        throw std::invalid_argument("something went wrong");
+    }
+
+    uint64_t fsize = static_cast<uint64_t>(fin.tellg());
+    fin.seekg(0, std::ios_base::beg);
+
+    T result;
+    char* encoded_data = new char[fsize];
+    fin.read(encoded_data, static_cast<std::streamsize>(fsize));
+    msgpack::unpack(encoded_data, fsize).get().convert(result);
+    return result;
+}
+
+// TODO(#7371) find a home for this
+acir_format::WitnessVector witness_map_to_witness_vector(std::map<std::string, std::string> const& witness_map)
```

Comment:
non-template functions inside a header should be marked 'inline' or they define competing symbols in linker if the header is included twice

Reply:
I consider this a blocking change

Reply:
Fixed